packages/lib/src/helpers/encodeCall.ts

```diff
@@ -1,15 +1,97 @@
+/*
+ * Utility that wraps around ethereumjs-abi, providing additional validation
+ * on passed type/value pair parameters.
+ */
+
 import abi from 'ethereumjs-abi';
+import * as util from 'ethereumjs-util';
 import BN from 'bignumber.js';
+import _ from 'lodash';
 
-function formatValue(value: any): string {
-  if (typeof(value) === 'number' || BN.isBigNumber(value)) return value.toString();
-  else if (typeof(value) === 'string' && value.match(/\d+(\.\d+)?e(\+)?\d+/)) return (new BN(value)).toString(10);
-  else return value;
-}
+const ERROR_MESSAGE_PREFIX = 'Encoding error';
+const ERROR_MESSAGE = (type: string, value: any) => `${ERROR_MESSAGE_PREFIX} for type ${type} and value ${value.toString()}`;
 
-export default function encodeCall(name: string, args: string[] = [], rawValues: any[] = []): string {
-  const values: string[] = rawValues.map(formatValue);
-  const methodId: string = abi.methodID(name, args).toString('hex');
-  const params: Buffer = abi.rawEncode(args, values).toString('hex');
+export default function encodeCall(name: string, types: string[] = [], rawValues: any[] = []): string {
+  if(types.length !== rawValues.length) throw new Error(ERROR_MESSAGE_PREFIX + '. Supplied number of types and values do not match.');
+  const values =_.zipWith(types, rawValues, (type: string, value: any) => parseTypeValuePair(type, value));
+  const methodId = abi.methodID(name, types).toString('hex');
+  const params = abi.rawEncode(types, values).toString('hex');
   return '0x' + methodId + params;
 }
+
+export function parseTypeValuePair(type: string, rawValue: any): any | never {
+  // Array type (recurse by calling this function with the individual elements).
+  if(/^[^\[]+\[.*\]$/.test(type)) { // Test for '[]' in type.
+    if(typeof rawValue === 'string') rawValue = rawValue.split(',');
+    if(rawValue.length === 0) return [];
+    const size = type.match(/(.*)\[(.*?)\]$/)[2]; // Find number between '[*]'.
+    if(size !== '' && parseInt(size, 10) !== rawValue.length) throw new Error(ERROR_MESSAGE(type, rawValue) + '. Invalid array length.');
+    const baseType = type.slice(0, type.lastIndexOf('[')); // Remove array part '[*]'.
+    return _.map(rawValue, (rawValueElement) => parseTypeValuePair(baseType, rawValueElement));
+  }
+  // Single type.
+  if(type === 'address') return parseAddress(type, rawValue);
+  else if(type === 'bool') return parseBool(type, rawValue);
+  else if(type === 'string') return parseString(type, rawValue);
+  else if(type.startsWith('bytes')) return parseBytes(type, rawValue);
+  else if(type.startsWith('uint')) return parseNumber(type, rawValue, true, true);
+  else if(type.startsWith('int')) return parseNumber(type, rawValue, false, true);
+  else if(type.startsWith('ufixed')) return parseNumber(type, rawValue, true, false);
+  else if(type.startsWith('fixed')) return parseNumber(type, rawValue, false, false);
+  else throw new Error(ERROR_MESSAGE(type, rawValue) + '. Unsupported or invalid type.');
```

Comment:
True! Given that it's a pretty rare usage of the ABI, do you think it would be ok to add a couple of TODOs here for now, and simply forward these types without erroring "Unsupported or invalid type"? Implementing a validation for these may be a bit complex. If you think that would be ok, I would create a new issue to deal with these more complex types in a separate PR.

Reply:
Absolutely, and we can delay that check until much later.

Reply:
Actually, I couldn't help myself and I implemented them :grimacing: 
Implementing the function type was trivial, and implementing tuples is almost identical to arrays.